src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala

```diff
@@ -92,20 +92,24 @@ trait InlineRewrite extends PredicateExpansion with InlineErrorChecker {
   }
 
   /**
-    * Removes given predicate unfolds and folds from statement.
+    * Transforms predicate unfolds and folds into assertions.
     *
     * @param stmts A Seqn whose statements will be traversed.
     * @param cond The condition a predicate must satisfy to no longer require (un)folding.
-    * @return The Seqn with all above unfolds and folds removed.
+    * @return The Seqn with all possible unfolds and folds turned into assertions.
     */
-  private[this] def removeFoldUnfolds(stmts: Seqn, cond: String => Boolean): Seqn = {
-    ViperStrategy.Slim({
-      case seqn@Seqn(ss, _) =>
-        seqn.copy(ss = ss.filterNot {
-          case Fold(PredicateAccessPredicate(PredicateAccess(_, name), _)) => cond(name)
-          case Unfold(PredicateAccessPredicate(PredicateAccess(_, name), _)) => cond(name)
-          case _ => false
-        })(seqn.pos, seqn.info, seqn.errT)
-    }, Traverse.TopDown).execute[Seqn](stmts)
+  private[this] def removeFoldUnfolds(stmts: Seqn, member: Member, program: Program, cond: String => Boolean): Seqn = {
+    ViperStrategy.CustomContext[Set[String]]({
+      case (fold@Fold(PredicateAccessPredicate(pred, perm)), ctxt) =>
+        if (cond(pred.predicateName)) {
+          val optPredbody = propagatePermission(pred.predicateBody(program, ctxt), perm)
+          (Assert(optPredbody.get)(fold.pos, fold.info, fold.errT), ctxt)
+        } else (fold, ctxt)
+      case (unfold@Unfold(PredicateAccessPredicate(pred, perm)), ctxt) =>
+        if (cond(pred.predicateName)) {
+          val optPredbody = propagatePermission(pred.predicateBody(program, ctxt), perm)
+          (Assert(optPredbody.get)(unfold.pos, unfold.info, unfold.errT), ctxt)
```

Comment:
Should be safe to call `.get` here right

Reply:
Yeah, since `cond` now checks that predicates are defined first

Reply:
Oh that wasn't a question
I do `.get` on line 111

Reply:
Oh It was a question, I'm just shitty at asking them online ty